autoload/lsp/lspserver.vim

```diff
@@ -792,19 +792,37 @@ enddef
 # LSP line and column numbers start from zero, whereas Vim line and column
 # numbers start from one. The LSP column number is the character index in the
 # line and not the byte index in the line.
-def GetLspPosition(): dict<number>
+def GetLspPosition(find_ident: bool): dict<number>
   var lnum: number = line('.') - 1
   var col: number = charcol('.') - 1
+  var line = getline('.')
+
+  if find_ident
+    # 1. skip to start of identifier
+    while line[col] != '' && line[col] !~ '\k'
+      col = col + 1
+    endwhile
+
+    # 2. back up to start of identifier
+    while col > 1 && line[col - 1] =~ '\k'
+      col = col - 1
+    endwhile
+
+    if line[col] == '' || col == 1 && line[col] !~ '\k'
```

Comment:
Note that charcol() returns the character number.  But the indexing of line[] is done here by using the byte index.  I am not sure whether this will cause a problem if multibyte characters are used in a keyword.

Reply:
Two things:

- Seems that in vim9 script `str[n]` works on characters, see `:h expr-[]`.

- Turns out that the character count should be based on utf-16 strings 🤕 https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocuments Making `😊` take up 2 characters instead of 1, or 4, which means this test will fail:

```
diff --git a/autoload/lsp/lspserver.vim b/autoload/lsp/lspserver.vim
index f0397c8..fcff366 100644
--- a/autoload/lsp/lspserver.vim
+++ b/autoload/lsp/lspserver.vim
@@ -795,6 +795,7 @@ enddef
 def GetLspPosition(): dict<number>
   var lnum: number = line('.') - 1
   var col: number = charcol('.') - 1
+  col += 4
   return {line: lnum, character: col}
 enddef
 
diff --git a/test/unit_tests.vim b/test/unit_tests.vim
index cbdf008..995f6a7 100644
--- a/test/unit_tests.vim
+++ b/test/unit_tests.vim
@@ -979,6 +979,40 @@ def Test_LspCustomNotificationHandlers()
   :%bw!
 enddef
 
+def Test_UnicodeColumnCalc()
+  :silent! edit Xtest.c
+  sleep 200m
+  var lines: list<string> =<< trim END
+    int count;
+    int fn(int a)
+    {
+      int 😊😊😊😊;
+      😊😊😊😊 = a;
+
+      int b;
+      b = a;
+      return    count + 1;
+    }
+  END
+  setline(1, lines)
+  :redraw!
+
+  # Test Unicode column calculation
+  cursor(5, 1) # 😊😊😊😊 = a;
+  search('a')
+  assert_equal([],
+              execute('LspGotoDefinition')->split("\n"))
+  assert_equal([2, 12], [line('.'), col('.')])
+
+  cursor(8, 1) # b = a;
+  search('a')
+  assert_equal([],
+              execute('LspGotoDefinition')->split("\n"))
+  assert_equal([2, 12], [line('.'), col('.')])
+
+  bw!
+enddef
+
```